pkg/pillar/cmd/volumemgr/blob.go

```diff
@@ -104,6 +116,195 @@ func downloadBlob(ctx *volumemgrContext, objType string, sv SignatureVerifier, b
+			if err := json.Unmarshal(data, &index); err != nil {
+				log.Errorf("loadBlobs(%s): could not parse index at %s: %+s", blob.Sha256, blobFile, err.Error())
+				return err
+			}
+			indexHash = sha
+		case types.BlobManifest:
+			// read it in so we can process it
+			data, err := ioutil.ReadAll(fileReader)
+			if err != nil {
+				log.Errorf("loadBlobs(%s): could not read data at %s: %+s", blob.Sha256, blobFile, err.Error())
+				return err
+			}
+			fileReader.Close()
+			// create a new reader for the content.WriteBlob
+			r = bytes.NewReader(data)
+			// try to parse the index
+			mfst := ocispec.Manifest{}
+			if err := json.Unmarshal(data, &mfst); err != nil {
+				log.Errorf("loadBlobs(%s): could not parse manifest at %s: %+s", blob.Sha256, blobFile, err.Error())
+				return err
+			}
+			manifests = append(manifests, &mfst)
+			manifestHashes = append(manifestHashes, sha)
+		default:
+			// do nothing special, just pass it on
+			r = fileReader
+		}
+
+		//Step 1.3: Ingest the blob into CAS
+		if err := ctx.casClient.IngestBlob(sha, r); err != nil {
+			err = fmt.Errorf("loadBlobs(%s): could not load blob file into containerd at %s: %+s",
+				blob.Sha256, blobFile, err.Error())
+			log.Errorf(err.Error())
+			return err
+		}
+		blob.IsLoaded = true
+		publishBlobStatus(ctx, blob)
+
+		//Step 1.4: If there is a VerifyImage reference, then we can remove that as we have successfully
+		// loaded the image into CAS
+		if blob.HasVerifierRef {
+			log.Infof("loadBlobs(%s): Blobs has verifyImage ref. removing that as the blob is loaded into CAS",
+				blob.Sha256)
+			MaybeRemoveVerifyImageConfig(ctx, blob.Sha256)
+			blob.HasVerifierRef = false
+			publishBlobStatus(ctx, blob)
+		}
+	}
+
+	//Step 2: Walk the tree from the root to add the necessary labels
+	if index != nil {
+		info := cas.BlobInfo{
+			Digest: indexHash,
+			Labels: map[string]string{},
+		}
+		for i, m := range index.Manifests {
+			info.Labels[fmt.Sprintf("%s.%d", containerdGCRef, i)] = m.Digest.String()
+		}
+		if err := ctx.casClient.UpdateBlobInfo(info); err != nil {
+			log.Errorf("loadBlobs(%s): could not update labels on index: %v", info.Digest, err.Error())
+			return err
+		}
+	}
+
+	if len(manifests) > 0 {
+		for j, m := range manifests {
+			info := cas.BlobInfo{
+				Digest: manifestHashes[j],
+				Labels: map[string]string{},
+			}
+			for i, l := range m.Layers {
+				info.Labels[fmt.Sprintf("%s.%d", containerdGCRef, i)] = l.Digest.String()
+			}
+			i := len(m.Layers)
+			info.Labels[fmt.Sprintf("%s.%d", containerdGCRef, i)] = m.Config.Digest.String()
+
+			if err := ctx.casClient.UpdateBlobInfo(info); err != nil {
+				log.Errorf("loadBlobs(%s): could not update labels on manifest: %v", info.Digest, err.Error())
+				return err
+			}
+		}
+
+	}
+	return nil
+}
+
+//MaybeAddRefToBlobs adds the refObject as an reference to the blobs in the give blob list.
+func MaybeAddRefToBlobs(ctx *volumemgrContext, refObjectID string, blobStatus ...*types.BlobStatus) {
```

Comment:
If `MaybeAddRefToBlobs` and `MaybeRemoveRefBlobStatus` are parallel (are they? did I misunderstand?) then they should have parallel names and parallel args (not one with pointers to a `BlobStatus` slide and the other to a bunch of hash strings).

Separately, why name it `MaybeAddRefToBlobs`, and not just `AddRefToBlobs`?

Reply:
FWIW the Maybe names we have elsewhere are of the form
MaybeAddAppNetworkConfig
which either creates an AppNetworkConfig, or if it exists, increases the refcount on it.

MaybeAddRefCount doesn't make sense as @deitch points out; either it is a MaybeAdd<Object> or a AddRefTo<object>

Reply:
I am not a stickler for the `Maybe*` naming, but I am for the parallel naming and args.

Reply:
Changed it to `AddRefToBlobStatus` and `RemoveRefFromBlobStatus`